lib/monitors/nvidiamonitor.cpp

```diff
@@ -10,163 +10,132 @@
 #include "nvidiamonitor.h"
 
 #include <algorithm>
-#include <errno.h>
-#include <fcntl.h>
 #include <iostream>
 #include <sstream>
-#include <string.h>
-
-#if defined(__linux__)
-#include <sys/wait.h>
-#include <unistd.h>
-#endif
+#include <memory>
+#include <array>
 
 const string errorMsg = "Failed to monitor nvidia-smi process";
 
-#if defined(__linux__)
-int read_nvidia_smi_stdout(int fd, string& gpuUsage, string& usedMem, string& totalMem)
-{
-    string line;
-    while (line.find('\n') == string::npos) { // full line read
-        char buffer[4096];
-        ssize_t count = read(fd, buffer, sizeof(buffer)); // if child process crashes, we gonna be blocked here forever
-        if (count == -1) {
-            return errno;
-        } else if (count > 0) { // there is something to read
-            line += string(buffer, count);
-        }
-    }
-
-    // Remove colon to have only spaces and use istringstream
-    auto noSpaceEnd = remove(line.begin(), line.end(), ',');
-    if (noSpaceEnd == line.end()) { // output trace does not have comma so something went wrong with the command
-        return ENODATA;
-    }
-
-    line.erase(noSpaceEnd, line.end());
-    std::istringstream ss(line);
-    ss >> gpuUsage >> usedMem >> totalMem;
-
-    return 0;
-}
-#endif
-
 uprofile::NvidiaMonitor::NvidiaMonitor()
 {
+    //query nvidia-smi to get the number of GPUs and initialize m_totalMem, m_usedMem, and m_gpuUsage vectors 
+    #if defined(__linux__)
+        try {
+            std::array<char, 128> buffer;
+            std::string result;
+            std::string cmd = "/usr/bin/nvidia-smi --query-gpu=count --format=csv,noheader,nounits";
+            std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
+            if (!pipe) {
+                throw std::runtime_error("popen() failed!");
+            }
+            while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
+                result += buffer.data();
+            }
+            nGPUs_ = static_cast<size_t>(std::stoull(result));
+            m_totalMem = std::vector<int>(nGPUs_, 0);
+            m_usedMem = std::vector<int>(nGPUs_, 0);
+            m_gpuUsage = std::vector<float>(nGPUs_, 0.0);
+        } catch (const std::exception& err) {
+            std::cerr << errorMsg << std::endl;
+        }
+    #endif
 }
 
 uprofile::NvidiaMonitor::~NvidiaMonitor()
-{
-    stop();
-}
-
-void uprofile::NvidiaMonitor::start(int period)
-{
-    watchGPU(period);
-}
+{}
+
+void uprofile::NvidiaMonitor::update_gpu_data(const std::vector<uprofile::NvidiaMonitor::Data>& data) {
```

Comment:
Calling each time nvidia-smi to retrieve the GPU info is more demanding than just starting nvidia-smi in a watching mode and listening the stdout through a dedicated thread. That's why I choose this solution when implementing NVidiaMonitor.

I profiled your implementation and each call of nvidia-smi takes roughly 60 to 75ms on my configuration. As it is called twice (one for memory, one for usage), it is less optimized than my initial version that was fetching every data in one single read.

Was there a limitation with the initial implementation that prevents you from fetching multiple GPUs?